Dockerfile

```diff
@@ -0,0 +1,55 @@
+# This Dockerfile is composed of two steps: the first one builds the release
+# binary, and then the binary is copied inside another, empty image.
+
+#################
+#  Build image  #
+#################
+
+FROM ubuntu:bionic AS build
+
+RUN apt-get update && DEBIAN_FRONTEND=noninteractive apt-get install -y \
+    ca-certificates \
+    curl \
+    build-essential \
+    pkg-config \
+    libssl-dev
+
+# Install the currently pinned toolchain with rustup
+COPY rust-toolchain /tmp/rust-toolchain
+RUN curl https://static.rust-lang.org/rustup/dist/x86_64-unknown-linux-gnu/rustup-init >/tmp/rustup-init && \
+    chmod +x /tmp/rustup-init && \
+    /tmp/rustup-init -y --no-modify-path --default-toolchain $(cat /tmp/rust-toolchain)
+ENV PATH=/root/.cargo/bin:$PATH
+
+# Build the dependencies in a separate step to avoid rebuilding all of them
+# every time the source code changes. This takes advantage of Docker's layer
+# caching, and it works by copying the Cargo.{toml,lock} with dummy source code
+# and doing a full build with it.
+WORKDIR /tmp/source
+COPY Cargo.lock Cargo.toml /tmp/source/
+COPY parser/Cargo.toml /tmp/source/parser/Cargo.toml
```

Comment:
Technically only the root Cargo.lock should matter I think.

Reply:
Unfortunately not. If I remove that line this is the output:

```
Step 9/18 : RUN cargo fetch
 ---> Running in 5ffccb66d836
error: failed to read `/tmp/source/parser/Cargo.toml`

Caused by:
  No such file or directory (os error 2)
```

Reply:
Oh, right, I was just thinking about caching -- not the actual requirements from Cargo.